### PR TITLE
Dropout support for memory efficient attention

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This PR builds off of:
- https://github.com/pytorch/pytorch/pull/101847
- https://github.com/pytorch/pytorch/pull/100583

It specifically adds dropout support to the memory efficient attention kernel. In the process of doing so roughly 3 changes were made:
- Update sdpa dispatching to allow for inputs requiring grad to be sent to efficient attention
- Update how memory efficient attention handles passing the rng state from forward to backward in order to enable cuda_graph support
- Fix a bug in the kernel that was causing incorrect gradients to be produced for num_keys > 64 with dropout and causal masking set. https://github.com/facebookresearch/xformers/pull/755

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov @bertmaher

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -2033,6 +2069,146 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
...
+                # High Precision Math Reference
+                out_ref = F.scaled_dot_product_attention(query_ref, key_ref, value_ref,
+                                                         dropout_p=dropout_p, is_causal=is_causal, scale=scale)
+                # Low Precision Math Reference
+                out_lp_ref = F.scaled_dot_product_attention(query_ref_lp, key_ref_lp, value_ref_lp,
+                                                            dropout_p=dropout_p, is_causal=is_causal, scale=scale)
+            else:
+                # Create the dropout_mask
+                dropout_mask = _get_mem_eff_drop_mask(batch_size, n_heads, seq_len_q, seq_len_k,
+                                                      dropout_p, ouput_seed, output_offset, device=device)
```
Comment (drisspg) [inline test/test_transformers.py:2164]: Explicitly convert to ints

File: aten/src/ATen/native/transformers/attention.cpp [hunk 2]
```diff
@@ -682,9 +682,9 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math(
     attn = at::softmax(attn, -1);
     if (dropout_p > 0.0) {
       if (dropout_mask.has_value()) {
-        auto attn_dropout_masked = attn.masked_fill(dropout_mask->logical_not(), 0.0);
+        attn = attn.masked_fill(dropout_mask->logical_not(), 0.0);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:685]: should have returned the dropout filled mask which makes it more useful for debugging

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 3]
```diff
@@ -720,22 +729,22 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_cuda(
       ? sdp::CustomMaskType::CausalFromTopLeft
       : sdp::CustomMaskType::NoCustomMask;
 
-  Tensor attention, log_sumexp;
-  std::tie(attention, log_sumexp) = at::_efficient_attention_forward(
+  Tensor attention, log_sumexp, seed, offset;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:732]: I think I can use structured bindings now

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 4]
```diff
@@ -1043,19 +1075,10 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
...
-  // uint64_t -> int64_t bitwise casting as PyTorch don't support uint64_t
-  // so just fake it as a int64_t
-  int64_t seed, offset;
-  if (use_dropout) {
-    std::memcpy(&seed, &rng_engine_inputs.seed_, sizeof(seed));
-    std::memcpy(&offset, &rng_engine_inputs.offset_.val, sizeof(offset));
-  }
-
-  return std::make_tuple(res, logsumexp);
+  return std::make_tuple(res, logsumexp, seed_t, offset_t);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1078]: std::move()

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 5]
```diff
@@ -1043,19 +1075,10 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
...
 #endif
   TORCH_CHECK(false, "USE_FLASH_ATTENTION was not enabled for build.")
-  return std::make_tuple(Tensor{}, Tensor{});
+  return std::make_tuple(Tensor{}, Tensor{}, Tensor{}, Tensor{});
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1081]: std::move()

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 6]
```diff
@@ -521,15 +529,18 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> _scaled_dot_product_flash_attenti
...
 
+
 std::tuple<at::Tensor, at::Tensor, at::Tensor> _scaled_dot_product_efficient_attention_backward_cuda(
     const at::Tensor& grad_out_,
     const at::Tensor& query,
     const at::Tensor& key,
     const at::Tensor& value,
     const at::Tensor& out,
     const at::Tensor& logsumexp,
+    const at::Tensor& rng_seed_tensor, // seed using for generating random numbers for dropout
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:540]: clean up comment

File: aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h [hunk 7]
```diff
@@ -156,6 +156,7 @@ struct AttentionKernel {
     int32_t head_dim_value;
     int32_t num_queries;
     int32_t num_keys;
+    int32_t num_keys_absolute;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h:159]: I should warp_uniform this right? @danthe3rd

**Review by drisspg — COMMENTED** [review]
closer

File: aten/src/ATen/native/native_functions.yaml [hunk 8]
```diff
@@ -14126,7 +14128,7 @@
     CUDA: _flash_attention_backward
 
 # Returns ouput, logsumexp if compute_logsumexp
-- func: _efficient_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int? max_seqlen_q, float dropout_p, int custom_mask_type, bool compute_log_sumexp=False, *, float? scale=None, Tensor? causal_diagonal=None, Tensor? seqlen_k=None) -> (Tensor output, Tensor logsumexp)
+- func: _efficient_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int? max_seqlen_q, float dropout_p, int custom_mask_type, bool compute_log_sumexp=False, *, float? scale=None, Tensor? causal_diagonal=None, Tensor? seqlen_k=None) -> (Tensor output, Tensor logsumexp, Tensor philox_seed, Tensor philox_offset)
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:14131]: nit: so are they `seed, offset`, `rng_seed_tensor, rng_offset_tensor`, or `philox_seed, philox_offset`? 3 functions, 3 versions

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:14131]: You are right, will make consistent

File: test/test_transformers.py [hunk 9]
```diff
@@ -2033,6 +2069,148 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
...
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA, "Does not support SDPA")
+    @parametrize("batch_size", [1, 8])
+    @parametrize("seq_len_q", [512, 1024])
+    @parametrize("seq_len_k", [512, 1024])
+    @parametrize("head_dim", [32, 64])
+    @parametrize("is_causal", [False, True])
+    @parametrize("dropout_p", [0.0, 0.22])
+    @parametrize("dtype", [torch.float16,])
+    @parametrize("scale", [None, "l1"])
+    def test_mem_efficient_attention_vs_math_ref_grads_cudagraph(self,
```
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py:2083]: Good call only thing that changes is the dropout call will update

File: test/test_transformers.py [hunk 10]
```diff
@@ -2033,6 +2069,148 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
...
+        # for the use of the fused kernel rather than the eager implemntation.
+        output_ref_atol, output_ref_rtol = get_tolerances(out_ref, out_lp_ref)
+
+        # Fudge Factor when dropout is enabled
+        dropout_fudge_factor = 1.0 if dropout_p == 0.0 else 1.5
+
+        query_fudge_factor = dropout_fudge_factor
+        grad_q_ref_atol, grad_q_ref_rtol = get_tolerances(query_ref.grad, query_ref_lp.grad, query_fudge_factor)
+
+        # TODO: Investigate why grad_k needs larger tolerances
```
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py:2198]: I debated copying this in, I kinda feel like it's a good breadcrumb for a future explorer lol

**Comment by ngimel** [comment]
Tests look good now, still inconsistency about names in native_functions

**Comment by desertfire** [comment]
> @desertfire I rebased this PR which was all green and is now failing on some new tests I suspect it is failing because the output has expanded to return the rng state and this test is comparing the rng state at successive runs.. any thoughts

I see. Do we expect the first output to be the same between eager and inductor runs? If yes, you can just return the first output here, https://github.com/pytorch/pytorch/blob/4bb2b65ea4912ec32d4106c122385aaad682b83e/test/inductor/test_torchinductor.py#L6315

File: aten/src/ATen/native/native_functions.yaml [hunk 11]
```diff
@@ -14101,12 +14101,14 @@
...
   dispatch:
     CUDA: _scaled_dot_product_efficient_attention_cuda
     NestedTensorCUDA: _scaled_dot_product_efficient_attention_nestedtensor_cuda
+  tags: nondeterministic_seeded
 
-- func: _scaled_dot_product_efficient_attention_backward(Tensor grad_out_, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, bool is_causal=False, bool chunk_grad_outputs=False, *, float? scale=None) -> (Tensor, Tensor, Tensor)
+- func: _scaled_dot_product_efficient_attention_backward(Tensor grad_out_, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, Tensor philox_seed, Tensor philox_offset, float dropout_p, bool is_causal=False, *, float? scale=None) -> (Tensor, Tensor, Tensor)
+  device_check: NoCheck
   dispatch:
     CUDA: _scaled_dot_product_efficient_attention_backward_cuda
```
Comment (cpuhrsch) [inline aten/src/ATen/native/native_functions.yaml:14113]: Wouldn't you want to add the `nondeterministic_seeded` tag here too?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 12]
```diff
@@ -763,23 +764,22 @@ _scaled_dot_product_efficient_attention_nestedtensor_cuda(
       ? sdp::CustomMaskType::CausalFromTopLeft
       : sdp::CustomMaskType::NoCustomMask;
 
-  Tensor attention, log_sumexp;
-  std::tie(attention, log_sumexp) = at::_efficient_attention_forward(
+  auto [attention, log_sumexp, seed, offset] = at::_efficient_attention_forward(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:767]: I'd add a small comment on why the return types for seed and offset have to be Tensors.

File: aten/src/ATen/native/transformers/attention.cpp (hunk 2 above)
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:685]: Wait doesn't this change the behavior?

```
auto attn_dropout_masked = attn.masked_fill(dropout_mask->logical_not(), 0.0);
[...]
return std::make_tuple(at::matmul(attn_dropout_masked, value * dropout_scaling), attn);
```

vs.

```
attn = attn.masked_fill(dropout_mask->logical_not(), 0.0);
[...]
return std::make_tuple(at::matmul(attn, value * dropout_scaling), *attn*);
```

You're using a different right hand operand in the second version. Is this fixing a bug?

Comment (cpuhrsch) [reply aten/src/ATen/native/transformers/attention.cpp:685]: Ah, never mind. dropout_mask is just a debug feature and you're updating it for testing.

File: aten/src/ATen/native/transformers/attention.cpp [hunk 13]
```diff
@@ -682,9 +682,9 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math(
     attn = at::softmax(attn, -1);
     if (dropout_p > 0.0) {
       if (dropout_mask.has_value()) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:684]: Maybe add a comment here that this isn't meant to be used by anyone directly? Or perhaps even raise a warning that this is for testing only?

Just in case someone attempts to use this in some other part of the codebase (however low the chances).

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 14]
```diff
@@ -875,18 +883,40 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
 
   at::Tensor res;
   at::Tensor logsumexp;
+  at::Tensor seed_t, offset_t;
 
   const bool use_dropout = std::fpclassify(dropout_p) != FP_ZERO;
-  at::PhiloxCudaState rng_engine_inputs;
+  at::PhiloxCudaState philox_state;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:889]: nit: I don't like renames for the sake of renames in large PRs, because I worry that I'll miss an important change as a reviewer and this PR already has a lot of changes.

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 15]
```diff
@@ -875,18 +883,40 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
...
+    // See Note [Acquire lock when using random generators]
     std::lock_guard<std::mutex> lock(gen->mutex_);
     // if using dropout, we produce 1 random number for each element of the
     // attention tensor
-    rng_engine_inputs = gen->philox_cuda_state(B * num_heads * M * N);
+    philox_state = gen->philox_cuda_state(B * num_heads * M * N);
+
+    if (at::cuda::currentStreamCaptureStatus() ==
+        at::cuda::CaptureStatus::None) {
+      auto [seed, offset] = at::cuda::philox::unpack(philox_state);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:902]: nit: I dislike auto for very low level, critical work. I'd rather this fail compilation than for the upcoming cast to succeed and introduce silent correctness.

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 16]
```diff
@@ -875,18 +883,40 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
...
+      seed_t = at::scalar_tensor(
+          at::Scalar(static_cast<int64_t>(seed)), at::dtype(at::kLong));
+      offset_t = at::scalar_tensor(
+          at::Scalar(static_cast<int64_t>(offset)), at::dtype(at::kLong));
+    } else {
+      seed_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+      offset_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+    }
+  } else {
+    if (at::cuda::currentStreamCaptureStatus() !=
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:912]: That "!=" vs. the "==" above makes this confusing to read.

Also why are you checking that state twice in a row?

You have
```
    if (at::cuda::currentStreamCaptureStatus() ==
        at::cuda::CaptureStatus::None) {
```
above
and then here you're doing
```
    if (at::cuda::currentStreamCaptureStatus() !=
        at::cuda::CaptureStatus::None) {
```

I'm probably missing something or you're not comparing against the right thing?

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 17]
```diff
@@ -1083,6 +1109,93 @@ bool _chunk_grad_outputs_efficient_attention(
...
+        mask_out[linear_idx] = (&rand_quad.x)[i];
+      }
+    }
+  }
+}
+} // namespace
+#endif
+/**
+ * fill tensor with random uniform values. only used for testing, not much
+ * attention is paid to performance
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1165]: heh

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 18]
```diff
@@ -211,7 +206,21 @@ _efficient_attention_backward(
...
-  at::PhiloxCudaState rng_engine_inputs(rng_seed, rng_offset);
+
+  at::PhiloxCudaState rng_engine_inputs;
+  if (use_dropout) {
+    if (at::cuda::currentStreamCaptureStatus() ==
+        at::cuda::CaptureStatus::None) {
+      rng_engine_inputs = at::PhiloxCudaState(
+          *philox_seed.data_ptr<int64_t>(),
+          *philox_offset.data_ptr<int64_t>());
+    } else { // dropout + capture
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:217]: I'd maybe expand on this comment a bit or tag something more central that explains this.

File: test/test_transformers.py [hunk 19]
```diff
@@ -1880,16 +1913,33 @@ def test_mem_efficient_attention_vs_math_ref_grads(self, device, batch_size: int
...
+                out_lp_ref = F.scaled_dot_product_attention(query_ref_lp, key_ref_lp, value_ref_lp,
+                                                            dropout_p=dropout_p, is_causal=is_causal, scale=scale)
+        else:
+            if seq_len_q > 1024:
+                self.skipTest("Will call _fill_mem_eff_dropout_mask with too many threads!")
+            # Create the dropout_mask
+            torch.manual_seed(seed)
+            dropout_mask = _get_mem_eff_drop_mask(batch_size, n_heads, seq_len_q, seq_len_k, dropout_p, seed, 0, device=device)
             # High Precision Math Reference
-            out_ref = F.scaled_dot_product_attention(query_ref, key_ref, value_ref,
```
Comment (cpuhrsch) [inline test/test_transformers.py:1888]: Why this change instead of running it under the guard with math set to True only?

File: aten/src/ATen/native/transformers/attention.cpp (hunk 13 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/attention.cpp:684]: great idea!

File: aten/src/ATen/native/transformers/cuda/attention.cu (hunk 16 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu:912]: No you are right this is weird

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu:912]: ophhh actually this is just word indentation

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu:912]: I still was able to clean it up

**Review by cpuhrsch — APPROVED** [review]

File: test/test_transformers.py (hunk 19 above)
Comment (drisspg) [reply test/test_transformers.py:1866]: the dropout_mask arg isn't exposed at the top level so call the math func directly
